### ISSUE TITLE
windows: CertificateBundleLoadFailure, most likely due to unrecognised system certificate

### ISSUE THREAD

**Issue by Deins** [ziglang/zig#15422]
### Zig Version

0.11.0-dev.2750+c4a63389e

### Steps to Reproduce and Observed Behavior

build zls for example on my windows 10 x64 machine:
```
PS W:\dev\zig\tools\zls> zig build
error: CertificateBundleLoadFailure 
```
I added err.logs to `std.crypto.certificate.zig` to find out from where it comes from and here  is trace:
```
error: CertificateBundleLoadFailure
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate.zig:682:40: 0x7ff7b5a03c11 in parseEnum__anon_248444 (zig.exe.obj)
    return E.map.get(oid_bytes) orelse {
                                       ^
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate.zig:657:5: 0x7ff7b578b595 in parseAlgorithm (zig.exe.obj)
    return parseEnum(Algorithm, bytes, element);
    ^
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate.zig:450:33: 0x7ff7b5541874 in parse (zig.exe.obj)
    const signature_algorithm = try parseAlgorithm(cert_bytes, algo_elem);
                                ^
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate\Bundle.zig:247:25: 0x7ff7b52085be in parseCert (zig.exe.obj)
    const parsed_cert = try Certificate.parse(.{
                        ^
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate\Bundle.zig:137:9: 0x7ff7b4defd04 in rescanWindows (zig.exe.obj)
        try cb.parseCert(gpa, decoded_start, now_sec);
        ^
W:\dev\zig\zig\stage3\lib\zig\std\crypto\Certificate\Bundle.zig:63:21: 0x7ff7b4c33194 in rescan (zig.exe.obj)
        .windows => return rescanWindows(cb, gpa),
                    ^
W:\dev\zig\zig\stage3\lib\zig\std\http\Client.zig:1002:61: 0x7ff7b4a80228 in request (zig.exe.obj)
            client.ca_bundle.rescan(client.allocator) catch return error.CertificateBundleLoadFailure;
                                                            ^
W:\dev\zig\zig\src\Package.zig:485:19: 0x7ff7b4a7e817 in fetchAndUnpack (zig.exe.obj)
        var req = try http_client.request(.GET, uri, h, .{});
                  ^
W:\dev\zig\zig\src\Package.zig:282:25: 0x7ff7b4a84443 in fetchAndAddDependencies (zig.exe.obj)
        const sub_pkg = try fetchAndUnpack(
                        ^
W:\dev\zig\zig\src\main.zig:4417:13: 0x7ff7b495d9b2 in cmdBuild (zig.exe.obj)
            try fetch_result;
            ^
W:\dev\zig\zig\src\main.zig:298:9: 0x7ff7b4903c1f in mainArgs (zig.exe.obj)
        return cmdBuild(gpa, arena, cmd_args);
        ^
W:\dev\zig\zig\src\main.zig:211:5: 0x7ff7b490149d in main (zig.exe.obj)
    return mainArgs(gpa, arena, args);
```
Also the unrecognized oid bytes are:
```
# std.log.err("oid_bytes: {any}", .{oid_bytes});
error: oid_bytes: { 43, 14, 3, 2, 29 }
```

So it seems that there is certificate installed on my system that is not recognized. 
I am not sure where to look up the oid bytes to find out what certificate is it. (please let me know where that can be looked up).

### Expected Behavior

For simple things like zig build I would expect it to warn but ignore unsupported certificates if possible.

**Comment by Deins** [comment]
So I managed to find, export and remove the certificate that was causing issues it seems: [UniversalADB](https://www.dropbox.com/s/nuof9aw0utxbxko/UniversalADB_1.cer?dl=0)

Confirmed that adding back specified certificate in `Thrusted root certificate authorities` makes the issue to happen again.
From inspecting it in with certificate manager GUI I don't see reason why it would cause problems. In text form its signature algorithm is `sha1RSA` is that different from `.sha1WithRSAEncryption` ?
![image](https://user-images.githubusercontent.com/8919956/233851422-4ab91458-0908-43c3-96db-d6af97470906.png)

**Comment by Deins** [comment]
Ok that certificate seems uses obsolete algorithm: http://oid-info.com/get/1.3.14.3.2.29. Probably not worth adding it to zig.
Also fyi found cool tool to preview the certificate contents in binary and parsed form: [online certificate parser](http://lapo.it/asn1js/#MIIC_DCCAeigAwIBAgIQms0ny9ciTKBIKOhff3H25zAJBgUrDgMCHQUAMBcxFTATBgNVBAMTDFVuaXZlcnNhbEFEQjAeFw0xOTA3MDcxNzA5NDFaFw0zOTEyMzEyMzU5NTlaMBcxFTATBgNVBAMTDFVuaXZlcnNhbEFEQjCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAKWgLeVXsTUmhVfaoJlSyX1EGzB6CWzj7iu59IUgj6yWISaJhMQioncRmBdUZFVRjnw-WNnPFo1ntTVhrlOTZSy0GRjFQxbqkAxrr3z62b-PQJ-2PdRMxoERv2sVsfWuYqLqQtavIaldkpYjJ9po0qtBv2I5U27PI80mPlqU4mJaz2548hnr1teywbUh5VrKNfTa0iZOiBHUHe4i6cdYmgYkeT9iS5qDighu7jR1I6od9tBc7Jojm6UQsX74egRqUxRTg12k-M1MinSrdBzOGzfxkFSx-Q0Xk-Guh3rgp9ki7XZ_39Q0ESyLeuO4tJslBGHjW02-BqBceF4-uj60Kh0CAwEAAaNMMEowSAYDVR0BBEEwP4AQhlfjtW30MxeOWHs8bXtGm6EZMBcxFTATBgNVBAMTDFVuaXZlcnNhbEFEQoIQms0ny9ciTKBIKOhff3H25zAJBgUrDgMCHQUAA4IBAQAFPPDcj3Sr_4flRvVpLRqtCgm_dRb37aohMNzdkxOi8jeGHrjia_Jp-I38afi12po7HiUcgBscvXu0hNGWCMgguM3mNHykluSzXN2wJ8jMQVMkUNdFJlipklTi6vMQIizyRjoiIV_OSHU9dFYwXZaQmz0L50cgWzFdfnX_Whh4-Jd5yeEPqeb2rtcpvRlmlnXZwmeOxhJLw9tUAnRVczjoQ2tw_JnXDiOjMwrHLdckQ7gG0sMM1Rwy3SsmPU0GuWR7Ei3XDrsq-lVJjXTvlWoQ9gIleRAmJHtgHhvDpnxhsEOJ7UlUARuq6WCGhnPhAgnjXezV1ELw7nYDmx6EJa-r)
If logging is allowed in std.crypto, would be useful to add some error reporting so that faulty certificates can be found easier:  https://github.com/Deins/zig/commit/f042421861668038370443640bb8eafe5e1b4cb6

**Comment by ERAGON007** [comment]
I'm having the same problem here with Zig version `0.11.0-dev.3000+d71a43ec2`

Please merge #15539 so we can know which certificate is causing the problem

**Comment by DraagrenKirneh** [comment]
> Please merge https://github.com/ziglang/zig/pull/15539 so we can know which certificate is causing the problem

That PR will not give you any information about which certificate that are failing. It will only ignore certificates which have an algorithm not supported in Zig.

@ERAGON007 
If you want to figure out which certificate is causing you trouble and you are on windows, then you can use this code: 

```zig
const std = @import("std");
const testing = std.testing;

const win = std.os.windows;

test "check store" {
    const alloc = std.testing.allocator;
    try checkCertStore(alloc);
}

fn checkCertStore(alloc: std.mem.Allocator) !void {
    std.debug.print("\n--- checking cert store --\n", .{});
    const w = std.os.windows;
    const GetLastError = w.kernel32.GetLastError;
    const root = [4:0]u16{ 'R', 'O', 'O', 'T' };
    const store = w.crypt32.CertOpenSystemStoreW(null, &root) orelse switch (GetLastError()) {
        .FILE_NOT_FOUND => return error.FileNotFound,
        else => |err| return w.unexpectedError(err),
    };
    defer _ = w.crypt32.CertCloseStore(store, 0);
    var ctx = w.crypt32.CertEnumCertificatesInStore(store, null);
    while (ctx) |context| : (ctx = w.crypt32.CertEnumCertificatesInStore(store, ctx)) {
        const encoded_cert = context.pbCertEncoded[0..context.cbCertEncoded];
        const info = @ptrCast(*Cert_Info, @alignCast(@alignOf(Cert_Info), context.pCertInfo));

        // std.debug.print("\nissuer:", .{});
        // try printBlob(alloc, &info.issuer);

        const certOrFail = std.crypto.Certificate.parse(.{
            .buffer = encoded_cert,
            .index = 0,
        });

        _ = certOrFail catch {
            std.debug.print("\nfailed to parse certificate ->\nissuer:", .{});
            try printBlob(alloc, &info.issuer);
            std.debug.print("subject:", .{});
            try printBlob(alloc, &info.subject);
        };
    }
    std.debug.print("\n\n --- checking done --- \n\n", .{});
}

const BitBlob = extern struct {
    count: win.DWORD,
    data: [*]win.BYTE,
    unused_bits: win.DWORD,
};

const PublicKeyInfo = extern struct {
    alg_id: SigAlg,
    public_key: BitBlob,
};

const Blob = extern struct {
    count: win.DWORD,
    data: [*]win.BYTE,

    pub fn slice(self: Blob) []const u8 {
        return self.data[0..self.count];
    }
};

const SigAlg = extern struct {
    oid: win.LPSTR,
    parameters: Blob,
};

const CertExtension = extern struct {
    oid: win.LPSTR,
    critical: win.BOOL,
    blob: Blob,
};

const Cert_Info = extern struct {
    version: win.DWORD,
    crypt_integer_blob: Blob,
    crypt_alg_id: SigAlg,
    issuer: Blob,
    not_before: win.FILETIME,
    not_after: win.FILETIME,
    subject: Blob,
    public_key_info: PublicKeyInfo,
    issuer_id: BitBlob,
    subject_id: BitBlob,
    extension: win.DWORD,
    cert_extension: *CertExtension,
};

pub extern "crypt32" fn CertNameToStrW(
    dwCertEncodingType: win.DWORD,
    pName: ?*Blob,
    dwStrType: win.DWORD,
    psz: ?win.LPSTR,
    csz: win.DWORD,
) callconv(win.WINAPI) win.DWORD;

fn printBlob(alloc: std.mem.Allocator, blob: *Blob) !void {
    var result = CertNameToStrW(0x1, blob, 1, null, 0);
    if (result == 0) return error.Fail;

    const ptr = try alloc.allocSentinel(u8, (result + 1) * 2, 0);
    defer alloc.free(ptr);
    result = CertNameToStrW(0x1, blob, 1, ptr, result);
    if (result == 0) return error.Fail2;

    std.debug.print("\n{s}\n", .{ptr[0..]});
}
```

**Comment by ERAGON007** [comment]
> That PR will not give you any information about which certificate that are failing. It will only ignore certificates which have an algorithm not supported in Zig.
> 
> @ERAGON007 If you want to figure out which certificate is causing you trouble and you are on windows, then you can use this code:

Alright, thanks for the info btw

**Comment by rohlem** [comment]
To me it reads like this should have been closed with the merge of #15539 (on May 22nd).

**Comment by whiter001** [comment]
0.14.0 Reproduce problem;
del all root with `certmgr.msc`
then run by admin 
`certutil -generateSSTFromWU roots.sst && certutil -addstore -f root roots.sst`
then run zig ok.
